File: pkg/java/src/main/java/dev/openfga/language/validation/DestructuredTupleToUserset.java
package dev.openfga.language.validation;

class DestructuredTupleToUserset {
    private final String decodedType;
    private final String decodedRelation;
    private final boolean wildcard;
    private final String decodedConditionName;


    public DestructuredTupleToUserset(String decodedType, String decodedRelation, boolean wildcard, String decodedConditionName) {
        this.decodedType = decodedType;
        this.decodedRelation = decodedRelation;
        this.wildcard = wildcard;
        this.decodedConditionName = decodedConditionName;
    }

    public String getDecodedType() {
        return decodedType;
    }

    public String getDecodedRelation() {
        return decodedRelation;
    }

    public boolean isWildcard() {
        return wildcard;
    }

    public String getDecodedConditionName() {
        return decodedConditionName;
    }

    public static DestructuredTupleToUserset from(String allowableType) {
        var tupleAndCondition = allowableType.split(" with ");
        var tupleString = tupleAndCondition[0];
        var decodedConditionName = tupleAndCondition.length > 1 ? tupleAndCondition[1] : null;
        var isWildcard = tupleString.contains(":*");
        var splittedWords = tupleString.replace(":*", "").split("#");
        return new DestructuredTupleToUserset(
                splittedWords[0],
                splittedWords.length > 1 ? splittedWords[1] : null,
                isWildcard,
                decodedConditionName);
    }

}

File: pkg/csharp/src/OpenFga.Language/Validation/DestructuredTupleToUserset.cs
namespace OpenFga.Language.Validation;

internal class DestructuredTupleToUserset
{
    public string DecodedType { get; }
    public string? DecodedRelation { get; }
    public bool IsWildcard { get; }
    public string? DecodedConditionName { get; }

    public DestructuredTupleToUserset(string decodedType, string? decodedRelation, bool isWildcard, string? decodedConditionName)
    {
        DecodedType = decodedType;
        DecodedRelation = decodedRelation;
        IsWildcard = isWildcard;
        DecodedConditionName = decodedConditionName;
    }

    public static DestructuredTupleToUserset From(string allowableType)
    {
        var tupleAndCondition = allowableType.Split(" with ");
        var tupleString = tupleAndCondition[0];
        var decodedConditionName = tupleAndCondition.Length > 1 ? tupleAndCondition[1] : null;
        var isWildcard = tupleString.Contains(":*");
        var splittedWords = tupleString.Replace(":*", "").Split('#');
        return new DestructuredTupleToUserset(
            splittedWords[0], 
            splittedWords.Length > 1 ? splittedWords[1] : null, 
            isWildcard, 
            decodedConditionName);
    }
}
